packages/pigeon/lib/cpp_generator.dart

```diff
@@ -1417,6 +1478,170 @@ void _writeSystemHeaderIncludeBlock(Indent indent, List<String> headers) {
   }
 }
 
+enum _FunctionOutputType { declaration, definition }
+
+/// Writes a function declaration or definition to [indent].
+///
+/// If [parameters] are given, each should be a string of the form 'type name'.
+void _writeFunction(
+  Indent indent,
+  _FunctionOutputType type, {
+  required String name,
+  String? returnType,
+  String? scope,
+  List<String> parameters = const <String>[],
+  List<String> startingAnnotations = const <String>[],
+  List<String> trailingAnnotations = const <String>[],
+  List<String> initializers = const <String>[],
+  void Function()? body,
+}) {
+  assert(body == null || type == _FunctionOutputType.definition);
+
+  // Set the initial indentation.
+  indent.write('');
+  // Write any starting annotations (e.g., 'static').
+  for (final String annotation in startingAnnotations) {
+    indent.add('$annotation ');
+  }
+  // Write the signature.
+  if (returnType != null) {
+    indent.add('$returnType ');
+  }
+  if (scope != null) {
+    indent.add('$scope::');
+  }
+  indent.add(name);
+  // Write the parameters.
+  if (parameters.isEmpty) {
+    indent.add('()');
+  } else if (parameters.length == 1) {
+    indent.add('(${parameters.first})');
+  } else {
+    indent.addScoped('(', null, () {
+      enumerate(parameters, (int index, final String param) {
```

Comment:
I did this the more verbose way instead of using `format` based on your recent feedback, since I think being verbose in this dedicated helper is much less problematic.